src/base/ERC721Permit.sol

```diff
@@ -0,0 +1,114 @@
+import {Address} from "@openzeppelin/contracts/utils/Address.sol";
+import {IERC1271} from "@openzeppelin/contracts/interfaces/IERC1271.sol";
+
+import {ChainId} from "../libraries/ChainId.sol";
+import {IERC721Permit} from "../interfaces/IERC721Permit.sol";
+
+/// @title ERC721 with permit
+/// @notice Nonfungible tokens that support an approve via signature, i.e. permit
+abstract contract ERC721Permit is ERC721, IERC721Permit {
+    mapping(address owner => mapping(uint256 word => uint256 bitmap)) public nonces;
+
+    /// @dev The hash of the name used in the permit signature verification
+    bytes32 private immutable nameHash;
+
+    /// @dev The hash of the version string used in the permit signature verification
+    bytes32 private immutable versionHash;
+
+    /// @notice Computes the nameHash and versionHash
+    constructor(string memory name_, string memory symbol_, string memory version_) ERC721(name_, symbol_) {
+        nameHash = keccak256(bytes(name_));
+        versionHash = keccak256(bytes(version_));
+    }
+
+    // TODO: implement here, or in posm
+    function tokenURI(uint256) public pure override returns (string memory) {
+        return "https://example.com";
+    }
+
+    /// @inheritdoc IERC721Permit
+    function DOMAIN_SEPARATOR() public view override returns (bytes32) {
+        return keccak256(
+            abi.encode(
+                // keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)')
+                0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f,
+                nameHash,
+                versionHash,
+                ChainId.get(),
+                address(this)
+            )
+        );
+    }
+
+    /// @inheritdoc IERC721Permit
+    /// @dev Value is equal to keccak256("Permit(address spender,uint256 tokenId,uint256 nonce,uint256 deadline)");
+    bytes32 public constant override PERMIT_TYPEHASH =
+        0x49ecf333e5b8c95c40fdafc95c1ad136e8914a8fb55e9dc8bb01eaa83a2df9ad;
+
+    /// @inheritdoc IERC721Permit
+    function permit(address spender, uint256 tokenId, uint256 deadline, uint256 nonce, uint8 v, bytes32 r, bytes32 s)
+        external
+        payable
+        override
+    {
+        require(block.timestamp <= deadline, "Permit expired");
+
+        address owner = ownerOf(tokenId);
+        require(spender != owner, "ERC721Permit: approval to current owner");
+
+        bytes32 digest = getDigest(spender, tokenId, nonce, deadline);
+
+        if (Address.isContract(owner)) {
+            require(IERC1271(owner).isValidSignature(digest, abi.encodePacked(r, s, v)) == 0x1626ba7e, "Unauthorized");
+        } else {
+            address recoveredAddress = ecrecover(digest, v, r, s);
+            require(recoveredAddress != address(0), "Invalid signature");
+            require(recoveredAddress == owner, "Unauthorized");
+        }
+
+        _useUnorderedNonce(owner, nonce);
+        approve(spender, tokenId);
```

Comment:
I think this line of code actually breaks the whole function.

When a user signs a permit, _anyone_ should be able to submit that permit signature. Otherwise they would just call `approve`. Like if Alice signs a permit, to allow bob to access her tokens, Bob should be able to submit the signature as long as he does so before the deadline. Or even charlie should be able to submit it on behalf of Bob, for Alice.

If you look at this function youre calling, it does the following:
```
        address owner = _ownerOf[id];
        require(msg.sender == owner || isApprovedForAll[owner][msg.sender], "NOT_AUTHORIZED");
        getApproved[id] = spender;
```

That means, that after all these signature checks, its THEN going to check if the caller is the owner or approved. But theyre not yet approved, and they shouldnt have to be the owner.

All of your tests use the helper function `LiquidityOperations._permit` , which has the line
```
_vm1.prank(signer);
```
before it submits the permit... ie you only test submitting a permit from the signer itself which is why your tests dont fail.

Pretty sure if you remove that prank, the tests will all start failing

Reply:
https://github.com/Uniswap/v4-periphery/issues/167